packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ImageCaptureHostApiImpl.java

```diff
@@ -0,0 +1,137 @@
+    if (flashMode != null) {
+      // This sets the requested flash mode, but may fail silently.
+      imageCaptureBuilder.setFlashMode(flashMode.intValue());
+    }
+    if (targetResolution != null) {
+      imageCaptureBuilder.setTargetResolution(
+          new Size(
+              targetResolution.getWidth().intValue(), targetResolution.getHeight().intValue()));
+    }
+    ImageCapture imageCapture = imageCaptureBuilder.build();
+    instanceManager.addDartCreatedInstance(imageCapture, identifier);
+  }
+
+  /** Sets the flash mode of the {@link ImageCapture} instance with the specified identifier. */
+  @Override
+  public void setFlashMode(@NonNull Long identifier, @NonNull Long flashMode) {
+    ImageCapture imageCapture =
+        (ImageCapture) Objects.requireNonNull(instanceManager.getInstance(identifier));
+    imageCapture.setFlashMode(flashMode.intValue());
+  }
+
+  /** Captures a still image and uses the result to return its absolute path in memory. */
+  @Override
+  public void takePicture(
+      @NonNull Long identifier, @NonNull GeneratedCameraXLibrary.Result<String> result) {
+    ImageCapture imageCapture =
+        (ImageCapture) Objects.requireNonNull(instanceManager.getInstance(identifier));
+    final File outputDir = context.getCacheDir();
+    File temporaryCaptureFile;
+    try {
+      temporaryCaptureFile = File.createTempFile("CAP", ".jpg", outputDir);
+    } catch (IOException | SecurityException e) {
+      handleImageFileError(result, "Cannot create file to save captured image: " + e.getMessage());
+      return;
+    }
+
+    ImageCapture.OutputFileOptions outputFileOptions =
+        cameraXProxy.createImageCaptureOutputFileOptions(temporaryCaptureFile);
+    ImageCapture.OnImageSavedCallback onImageSavedCallback =
+        createOnImageSavedCallback(temporaryCaptureFile, result);
+
+    imageCapture.takePicture(
+        outputFileOptions, Executors.newSingleThreadExecutor(), onImageSavedCallback);
+  }
+
+  /** Creates a callback used when saving a captured image. */
+  @VisibleForTesting
+  public ImageCapture.OnImageSavedCallback createOnImageSavedCallback(
+      @NonNull File file, @NonNull GeneratedCameraXLibrary.Result<String> result) {
+    return new ImageCapture.OnImageSavedCallback() {
+      @Override
+      public void onImageSaved(@NonNull ImageCapture.OutputFileResults outputFileResults) {
+        result.success(file.getAbsolutePath());
+      }
+
+      @Override
+      public void onError(@NonNull ImageCaptureException exception) {
+        handleImageFileError(result, getOnImageSavedExceptionDescription(exception));
+      }
+    };
+  }
+
+  /** Handle errors with creating a file to save captured image or capturing an image. */
+  private void handleImageFileError(
+      @NonNull GeneratedCameraXLibrary.Result<String> result, @NonNull String errorDescription) {
+    // Send empty path because image was not saved.
+    result.success("");
```

Comment:
Looking particularly for feedback on using an empty path to signify a file error.

Reply:
What are the other options? In general I would expect either a typed result with meta data or thrown exception to handle (in java). If there was going to be a string return value and neither of the others then I would expect null not empty.

Reply:
I agree with you. I think I got caught up in the confusion of resulting with an error versus sending a camera error to the Dart stream. It's my understanding now (based on the current plugin) that we should send a camera error to the stream if an error occurs while an operation is ongoing. Otherwise, we can throw the error directly from the native side. I changed it to that [here](https://github.com/flutter/packages/pull/3287/files#diff-86d57fcd214e52f715522590f682b0bf51c7c56caa13384fbec553504a772777R92).

Let me know what you think. Also @bparrishMines will this work well with the wrapping pattern? I want to make sure we throw errors consistently from here on out.

Reply:
Yea I agree that this is the ideal way to handle it and is how we handle it for the `webview_flutter` implementations as well. The native code can just return the exception in `result.error` and then the Dart platform implementation can handle converting it to a `CameraException`.